### PR TITLE
Defer to database validation when default values for mandatory stored procedure params are excluded from config

### LINKED ISSUE #1748: [Bug]: Optional stored procedure parameters generate the error "Missing required procedure parameters"
### What happened?

I have a stored procedure like the following:

```sql
create procedure dbo.stp_Dummy
@mandatoryParam int,
@optionalParam int = 10
as
select 
    @mandatoryParam as mandatoryParam,
    @optionalParam as optionalParam
go  
```

for which the `@optionalParam` is, as the name implies, optional. 

If I configure DAB to use that stored procedure using the following configuration, omitting the `optionalParam` (since it is optional):

```json
"Dummy": {
    "source": {
      "type": "stored-procedure",
      "object": "dbo.stp_Dummy",
      "parameters": {
        "mandatoryParam": 10
      }
    },
    "graphql": {
      "operation": "query"
    },
    "permissions": [
      {
        "role": "anonymous",
        "actions": [
          "execute"
        ]
      }
    ]
  }
```

I get the following error during the startup:

```text
 Azure.DataApiBuilder.Service.Exceptions.DataApiBuilderException: Invalid request. Missing required procedure parameters: optionalParam for entity: Dummy
```

### Version

Microsoft.DataApiBuilder 0.8.52+c69925060e1942d28515b9c4b89ea24832da0c7c

### What database are you using?

Azure SQL

### What hosting model are you using?

Local (including CLI)

### Which API approach are you accessing DAB through?

REST, GraphQL

### Relevant log output

```Text
Information: Microsoft.DataApiBuilder 0.8.52+c69925060e1942d28515b9c4b89ea24832da0c7c
Information: Config not provided. Trying to get default config based on DAB_ENVIRONMENT...
Information: Environment variable DAB_ENVIRONMENT is (null)
Loading config file from dab-config.json.
Information: Loaded config file: dab-config.json
Information: Setting default minimum LogLevel: Debug for Development mode.
Starting the runtime engine...
Loading config file from dab-config.json.
info: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[63]
      User profile is available. Using 'C:\Users\damauri\AppData\Local\ASP.NET\DataProtection-Keys' as key repository and Windows DPAPI to encrypt keys at rest.
info: Azure.DataApiBuilder.Core.Services.ISqlMetadataProvider[0]
      Dummy path: /api/Dummy
dbug: Azure.DataApiBuilder.Core.Resolvers.IQueryExecutor[0]
      Executing query:
SELECT name as result_field_name, TYPE_NAME(system_type_id) as result_type, is_nullable FROM sys.dm_exec_describe_first_result_set_for_object (OBJECT_ID('dbo.stp_Dummy'), 0) WHERE is_hidden is not NULL AND is_hidden = 0
dbug: Azure.DataApiBuilder.Core.Services.ISqlMetadataProvider[0]
      Logging primary key information for entity: Dummy.
info: Azure.DataApiBuilder.Core.Configurations.RuntimeConfigValidator[0]
      Validating Relationship Section in Config...
fail: Azure.DataApiBuilder.Service.Startup[0]
      Unable to complete runtime initialization. Refer to exception for error details.
      Azure.DataApiBuilder.Service.Exceptions.DataApiBuilderException: Invalid request. Missing required procedure parameters: optionalParam for entity: Dummy
         at Azure.DataApiBuilder.Core.Configurations.RuntimeConfigValidator.ValidateStoredProceduresInConfig(RuntimeConfig runtimeConfig, ISqlMetadataProvider sqlMetadataProvider)
         at Azure.DataApiBuilder.Service.Startup.PerformOnConfigChangeAsync(IApplicationBuilder app)
fail: Azure.DataApiBuilder.Service.Startup[0]
      Exiting the runtime engine...
crit: Microsoft.AspNetCore.Hosting.Diagnostics[6]
      Application startup exception
      System.ApplicationException: Could not initialize the engine with the runtime config file: dab-config.json
         at Azure.DataApiBuilder.Service.Startup.Configure(IApplicationBuilder app, IWebHostEnvironment env, RuntimeConfigProvider runtimeConfigProvider)
         at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
         at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
         at Microsoft.AspNetCore.Hosting.ConfigureBuilder.Invoke(Object instance, IApplicationBuilder builder)
         at Microsoft.AspNetCore.Hosting.ConfigureBuilder.<>c__DisplayClass4_0.<Build>b__0(IApplicationBuilder builder)
         at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass15_0.<UseStartup>b__1(IApplicationBuilder app)  
         at Microsoft.AspNetCore.Mvc.Filters.MiddlewareFilterBuilderStartupFilter.<>c__DisplayClass0_0.<Configure>g__MiddlewareFilterBuilder|0(IApplicationBuilder builder)
         at Microsoft.AspNetCore.HostFilteringStartupFilter.<>c__DisplayClass0_0.<Configure>b__0(IApplicationBuilder app)       
         at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
Unable to launch the runtime due to: System.ApplicationException: Could not initialize the engine with the runtime config file: dab-config.json
   at Azure.DataApiBuilder.Service.Startup.Configure(IApplicationBuilder app, IWebHostEnvironment env, RuntimeConfigProvider runtimeConfigProvider)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.Invoke(Object instance, IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.<>c__DisplayClass4_0.<Build>b__0(IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass15_0.<UseStartup>b__1(IApplicationBuilder app)        
   at Microsoft.AspNetCore.Mvc.Filters.MiddlewareFilterBuilderStartupFilter.<>c__DisplayClass0_0.<Configure>g__MiddlewareFilterBuilder|0(IApplicationBuilder builder)
   at Microsoft.AspNetCore.HostFilteringStartupFilter.<>c__DisplayClass0_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Azure.DataApiBuilder.Service.Program.StartEngine(String[] args)
Error: Failed to start the engine.
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by seantleonard**
## Why make this change?

- Closes #1748
- DAB should not crash during startup for the following scenario:
  1. Database Stored procedure defined with 2 parameters (at least 1 mandatory param). DAB config defines default values for 1 of 2 of those parameters. Consequently, DAB fails config validation because the number of default parameters defined does not match the number of parameters in the database. However, this is not an issue when 0 config param defaults are defined. This leads to the next scenario:
  2. DAB should not fail a request when config doesn't define default values for parameters and database does. Should let database handle the error because DAB has no way of resolving mandatory vs. optional stored procedure parameters.

## What is this change?

1. Adds new sub-status code `DatabaseInputError` that is used when MSSQL returns error code 201 with message:
> Procedure or function '%.*ls' expects parameter '%.*ls', which was not supplied.
1. With this change, DAB **now returns** the following error to the client when DAB config doesn't define default values for mandatory stored procedure parameters:
    > Invalid request. Missing required request parameters.

     1. The db returned error contains database object metadata which we have typically avoided returning to the client from DAB engine because the dab config file typically overrides any database object names. DAB will return
1. `DabGraphQLResultSerializer` class added which ensures that responses indicating 'missing procedure parameters' result in HTTP 400 for the GraphQL endpoint. REST endpoint returns 400 without needing the result serializer.

### Example requests
Use the sample db schema and dab config from the linked issue.
1. GraphQL Endpoint
```graphql
query ExecuteDummy {
    executeDummy {
        mandatoryParam
        optionalParam
    }
}
```
Response HTTP 400
```json
{
  "errors": [
    {
      "message": "Invalid request. Missing required request parameters.",
      "extensions": {
        "message": "Invalid request. Missing required request parameters.",
        "stackTrace": "<stack trace omitted for brevity>",
        "code": "DatabaseInputError"
      }
    }
  ]
}
```

3. REST endpoint
```https
GET localhost:5000/api/Dummy?optionalParam=2345
```

Response HTTP 400
```json
{
    "error": {
        "code": "DatabaseInputError",
        "message": "Invalid request. Missing required request parameters.",
        "status": 400
    }
}
```

## How was this tested?
updates stored procedure tests which check for missing required request parameters. 
- [x] Integration Tests
- [x] Unit Tests

### CODE REVIEW

**Comment by Aniruddh25** [comment]
> DAB returns the following error to the client when DAB config doesn't define default values for mandatory stored procedure parameters:
> 
> Invalid request. Missing required procedure parameters.
> 
> The db returned error contains database object metadata which we have typically avoided returning to the client from DAB engine because the dab config file typically overrides any database object names. DAB will return

I dont follow the placement of this in the description under `What is the change?` section.

**Comment by seantleonard** [comment]
> DAB returns

@Aniruddh25 I missed a word. Now reads: "With this change, DAB **now** returns..."

File: src/Core/Services/DabGraphQLResultSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/// - DatabaseInputError. This indicates that the client can make a change to request contents to influence
+/// a change in the response.
+/// </summary>
+public class DabGraphQLResultSerializer : DefaultHttpResultSerializer
+{
+    public override HttpStatusCode GetStatusCode(IExecutionResult result)
+    {
+        if (result is IQueryResult queryResult && queryResult.Errors?.Count > 0)
+        {
+            if (queryResult.Errors.Any(error => error.Code == DataApiBuilderException.SubStatusCodes.DatabaseInputError.ToString()))
```
Comment (Aniruddh25) [inline src/Core/Services/DabGraphQLResultSerializer.cs:25]: Is DatabaseInputError is the only error that will be classified as a BadRequest? Are there any errors that can be classified as such?

**Review by Aniruddh25 — APPROVED** [review]
LGTM, a few questions and nits

File: src/Core/Resolvers/Sql Query Structures/SqlExecuteQueryStructure.cs [hunk 2]
```diff
@@ -43,35 +41,31 @@ public SqlExecuteStructure(
...
                 }
                 else
                 {
-                    // Fill with default value from runtime config
+                    // When the runtime config defines a default value for a parameter,
+                    // create a parameter using that value and add it to the query structure.
+                    // Database metadata does not indicate whether a SP parameter has a default value
+                    // and as a result, we don't know whether an SP parameter is optional.
+                    // Therefore, DAB relies on the database error to indicate that a required parameter is missing.
                     if (paramDefinition.HasConfigDefault)
```
[the remark this answers is not in the document]
Comment (seantleonard) [reply src/Core/Resolvers/Sql Query Structures/SqlExecuteQueryStructure.cs]: yep, revised. Looks cleaner now

File: src/Core/Resolvers/MsSqlDbExceptionParser.cs [hunk 3]
```diff
@@ -56,6 +59,8 @@ public MsSqlDbExceptionParser(RuntimeConfigProvider configProvider) : base(confi
                 "548", "2627", "22818", "22819", "22820", "22821",
                 "22822", "22823", "22824", "22825", "3960", "5062"
             });
+
+            _errorMessages.Add(201, "Invalid request. Missing required procedure parameters.");
```
[the remark this answers is not in the document]
Comment (seantleonard) [reply src/Core/Resolvers/MsSqlDbExceptionParser.cs]: Yeah that's fine. I've updated the error message to reflect that.

File: src/Core/Resolvers/DbExceptionParser.cs [hunk 4]
```diff
@@ -65,5 +76,25 @@ public Exception Parse(DbException e)
...
         public abstract HttpStatusCode GetHttpStatusCodeForException(DbException e);
+
+        /// <summary>
+        /// Gets a specific substatus code which describes the cause of the error in more detail.
+        /// </summary>
+        /// <param name="e">The exception thrown as a result of execution of the request.</param>
+        /// <returns>status code to be returned in the response.</returns>
+        public virtual SubStatusCodes GetResultSubStatusCodeForException(DbException e)
+        {
+            return SubStatusCodes.DatabaseOperationFailed;
```
[the remark this answers is not in the document]
Comment (seantleonard) [reply src/Core/Resolvers/DbExceptionParser.cs:88]: updated

File: src/Core/Services/DabGraphQLResultSerializer.cs (hunk 1 above)
Comment (seantleonard) [reply src/Core/Services/DabGraphQLResultSerializer.cs:25]: There may be. We will need to go through remaining error codes to see which ones we want to return as HTTP 400. This pr was specifically scoped for Stored Proc's bad request message.

**Comment by abhishekkumams** [comment]
would it be possible to add the missing required params, in the error message?

File: src/Service.Tests/SqlTests/RestApiTests/Find/FindApiTestBase.cs [hunk 5]
```diff
@@ -1476,8 +1476,10 @@ await SetupAndRunRestApiTest(
         }
 
         /// <summary>
-        /// Validates that parameters of a SP sent through the request body are ignored and
-        /// results in a 400 Bad Request due to missing required parameters. 
+        /// For a HTTP GET request, validates that request body contents are ignore.
```
Comment (abhishekkumams) [inline src/Service.Tests/SqlTests/RestApiTests/Find/FindApiTestBase.cs:1479]:
```suggestion
        /// For a HTTP GET request, validates that request body contents are ignored.
```

File: src/Service.Tests/SqlTests/RestApiTests/Find/FindApiTestBase.cs [hunk 6]
```diff
@@ -1476,8 +1476,10 @@ await SetupAndRunRestApiTest(
...
+        /// e.g. Any stored procedure parameters sent through the request body
+        /// are not considered.
+        /// Validates that a request whicResults in a 400 Bad Request due to missing required parameters.
```
Comment (abhishekkumams) [inline src/Service.Tests/SqlTests/RestApiTests/Find/FindApiTestBase.cs:1482]:
```suggestion
        /// Validates that a request which results in a 400 Bad Request due to missing required parameters. 
```

**Review by abhishekkumams — APPROVED** [review]

**Comment by seantleonard** [comment]
> would it be possible to add the missing required params, in the error message?

Yes, for dev mode, we can return the fields from the stored procedure. Unfortunately, dab currently doesn't have mechanism in place to perform request validation after result has been returned from db. Determining the mapped field names which were excluded from the request is not possible here. That would be a good enhancement to have at some point if there is demand for it.

**Review by Aniruddh25 — APPROVED** [review]

**Pull request merged**